### DISCUSSION TITLE
io_uring Support

### DISCUSSION THREAD

**Discussion by g-tejas** [eclipse-ecal/ecal#1128] [Ideas]
Is there any planned support for [io_uring](https://github.com/axboe/liburing) in ecal?

**Comment by FlorianReimold** [comment]
Since this is the first time I have heard about that, probably not ;)

But we can start a discussion about it! What should we use it for? What problem do you think this would solve in eCAL?

**Comment by g-tejas** [comment]
I'm still learning about this, but from what I understand, io_uring (which uses the liburing library) is a new way for Linux to handle I/O operations. It uses two ring buffers to communicate between the kernel and user space, and allows batch processing of multiple I/O operations with a single system call. This could make things faster and more efficient compared to older methods like epoll and kqueue.

In terms of networking and storage, io_uring might help speed things up and prevent the application from stalling during read and write operations. So, it seems like adding io_uring to eCAL could improve its performance in situations where there's a lot of I/O happening.

But take this with a grain of salt, I'm not an expert! You can read more about io_uring on this blog: https://unixism.net/loti/what_is_io_uring.html#what-is-io-uring

**Reply by FlorianReimold** [reply]
Thanks for the overview. We are currently using asio for network operations. As far as I read, asio currently does not use io_uring.

Btw: Do you experience performance issues in eCAL that you would expect io_uring to solve?

**Reply by KristoferHansson** [reply]
asio has had support for io_uring since [late 2021](https://github.com/chriskohlhoff/asio/commit/36440a92eb83da34b7516af2632b119f83b66a35), in version 1.22.
As for io_uring and performance, disc I/O has been proven, networking I/O seems to be not that obvious(yet).
